text/0000-cargo-artifact-specific-dependencies.md

```diff
@@ -0,0 +1,121 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A single project may include multiple artifacts - zero or one libraries, and any number of binaries, like the following `Cargo.toml`:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
+
+[lib]
+name = "myproject"
+path = "src/lib.rs"
+
+[[bin]]
+name = "myproject_cli"
+path = "src/main.rs"
+```
+
+This project contains a library usable by other crates, but it also includes a command-line interface to allow users to invoke features of the library from the command line.
+
+To streamline the user experience, we may want to use an argument parsing library, lke `clap` or `structopt`. However, adding it as a dependency would add it for all artifacts - including the library, which doesn't use it at all.
+
+To solve this, we can add dependencies for a specific artifact, like so:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
+
+[lib]
+name = "myproject"
+path = "src/lib.rs"
+
+[[bin]]
+name = "myproject_cli"
+path = "src/main.rs"
+dependencies = { clap = "*" }
+```
+
+Now, `clap` is required for the binary `myproject_cli`, but not for the library `myproject`, and can be used in the same way as crate-wide dependencies. In this example, `clap` could be used from the binary artifact, but attempting to use it from the library artifact would result in an unresolved import error. This has the potential to improve compile time and reduce binary size by omitting dependencies where they're not actually required. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature would add a new `dependencies` key to the `bin`, `lib`, `test`, `bench`, and `example` sections of the `Cargo.toml` file, accepting a table describing dependencies that are used only for that artifact, in addition to the ones defined in the top-level `dependencies` and `dev-dependencies` sections. These dependencies would be specified in the same way they would be in the top-level `dependencies` section, allowing version patterns, features, and alternative sources to be specified [as described in the Cargo manual](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).
+
+Additionally, artifact-specific dependencies can be specified as a non-inline table. This is semantically identical to declaring the dependencies inline, and the syntax would look similar to crate-wide dependencies:
+
+```toml
+[[bin]]
+name = "myproject_cli"
+
+[bin.myproject_cli.dependencies]
```

Comment:
Using just `[bin.dependencies]` here will result in an identical TOML document as the inline version, avoiding an extra level of nesting in the output, see [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=574a94a190a1ae8c00146d8fe7bc1a79).

Reply:
That's interesting; good point. You could even just use `[dependencies]`... though I'm not sure if that's more confusing, given the existing top-level table of that name?

Reply:
Using `[dependencies]` would refer to the top-level table `dependencies`, so you'll get `ErrorKind::DuplicateTable("dependencies")`. The array-of-tables syntax specifically supports this case of having a table-path prefixed with an array-of-tables path as being a sub-element of the last table in that array.

Reply:
Oh, you're right, sorry. I was misreading the playground output. This sounds good.

Reply:
If you fancy updating the RFC text, @dmarcuse, I would probably just do that. (Do you agree this makes sense?)

Reply:
That makes sense to me, I wasn't aware of that syntax feature. I've updated the RFC to reflect this!